lib/diff.mli

```diff
@@ -1,100 +1,82 @@
 
-type 'same atomic_option =
-  ('same, 'same atomic_modification) option_
+type ('same, 'diff) option_ = ('same option, ('same, 'diff) entry) maybe_changed
+type 'same atomic_option = ('same, 'same atomic_modification) option_
 
 type ('same, 'diff) atomic_variant =
   ('same, ('same atomic_maybe_changed, 'diff) maybe_changed) maybe_changed
 
 type ('same, 'diff) list_ = ('same list, 'diff list) maybe_changed
 
-type value = {
-  vname : string;
-  vdiff : Types.value_description atomic_t;
-}
-
-type type_ = {
-  tname : string;
-  tdiff : (Types.type_declaration, type_modification) t;
-}
-
 and type_modification = {
   type_kind : (Types.type_decl_kind, type_kind) atomic_variant;
   type_privacy : (Asttypes.private_flag, type_privacy) maybe_changed;
   type_manifest : Types.type_expr atomic_option;
-  type_params: (Types.type_expr, type_param) list_
+  type_params : (Types.type_expr, type_param) list_;
 }
 
-and type_param = (Types.type_expr, type_param_diff) maybe_changed
-
-and type_param_diff =
-  | Added_tp of Types.type_expr
-  | Removed_tp of Types.type_expr
-
-and type_privacy = Added_p | Removed_p
-
 and type_kind =
-  | Record_tk of record_field list
-  | Variant_tk of constructor_ list
+  | Record_tk of (Types.label_declaration, record_field) map
+  | Variant_tk of
+      ( Types.constructor_declaration,
+        (Types.constructor_declaration, cstr_args) atomic_variant )
+      map
 
 and record_field = {
-  rname : string;
-  rdiff : Types.label_declaration atomic_t;
+  type_expr : Types.type_expr atomic_maybe_changed;
+  mutable_flag : (Asttypes.mutable_flag, field_mutability) maybe_changed;
 }
 
-and constructor_ = {
-  csname : string;
-  csdiff : (Types.constructor_declaration, constructor_modification) atomic_variant;
-}
+and field_mutability = Added_m | Removed_m
 
-and constructor_modification =
-  | Record_c of record_field list
-  | Tuple_c of tuple_component list
+and cstr_args =
+  | Record_c of (Types.label_declaration, record_field) map
+  | Tuple_c of
+      (Types.type_expr, Types.type_expr atomic_maybe_changed_entry) list_
 
-and tuple_component = Types.type_expr atomic_maybe_changed
+and type_privacy = Added_p | Removed_p
 
-type class_ = {
-  cname : string;
-  cdiff : Types.class_declaration atomic_t;
-}
+and type_param = (Types.type_expr, type_param_diff) maybe_changed
 
-and cltype = {
-  ctname : string;
-  ctdiff : Types.class_type_declaration atomic_t;
-}
+and type_param_diff =
+  | Added_tp of Types.type_expr
+  | Removed_tp of Types.type_expr
 
 type module_ = {
   mname : string;
-  mdiff : (Types.module_declaration, signature_modification) t;
-}
-
-and modtype = {
-  mtname : string;
-  mtdiff : (Types.modtype_declaration, signature_modification) t;
+  mdiff : (Types.module_declaration, signature_modification) entry;
 }
 
-and signature_modification = Unsupported | Supported of sig_item list
+and signature_modification =
+  | Unsupported
+  | Supported of (Types.signature_item, sig_item) map
```

Comment:
Unfortunately this has either to be several maps, or not to be "string_maps": the namespace for values and types are not the same, so as it is written there could be clashes.